link/Elf.zig: ensure capacity before appending linker args.

xxxbxxx opened the pull request:
fixes e567abb339e1edaf5a3c86fe632522a3b8005275 "rework linker inputs" 

closes https://github.com/ziglang/zig/issues/21801

alexrp reviewed (approved):
Fix looks fine in principle.

However, given that we can't really predict how many arguments we're *actually* going to append, and considering that populating `argv` is not at all performance-sensitive, I do kind of question the value of using `argv.appendAssumeCapacity()` here... as opposed to just doing `argv.append()` like the rest of this entire function is doing, including the `-whole-archive` handling (which is a bit similar). Just doesn't seem worth the risk of bugs like the one we ran into here. :shrug:

(To be clear, I'm referring to the `argv.appendAssumeCapacity()` calls below, not the one added in this PR.)

alexrp commented:
In any case, I've confirmed locally that this fixes the issue, so will merge as-is to unbreak master.

xxxbxxx commented:
> I do kind of question the value of using `argv.appendAssumeCapacity()` here... as opposed to just doing `argv.append()`

yeah I wasn't sure. 

at first I removed "AssumeCapacity", but then I changed my mind thinking that although it would be preposterous to account for all the potential "append" in the function and alloc once, 
this was a reasonably small and scoped part, and  it was easy enough to do "the right thing." 
whatever.
